spec/app/ics-028-cross-chain-valset-update/README.md

```diff
@@ -0,0 +1,257 @@
+
+Note that the protocol we present is generalized to multiple baby blockchains.
+In other words, a single parent blockchain might have multiple baby blockchains under its "jurisdiction".
+Hence, the Cross-Chain Validation protocol is concerned with more than two entities (one parent blockchain and potentially multiple baby blockchains).
+
+### Properties
+
+This subsection is devoted to defining properties that the Cross-Chain Validation protocol ensures.
+Recall that the parent blockchain has an ability to demand a change to the validator set of the baby chain.
+Moreover, we want to ensure the stake of validators of the baby blockchain are "frozen" at the parent chain.
+
+We present the interface of the protocol:
+  - Request: <ChangeValidatorSet, babyChain, V> - request made by the parent blockchain to change the validator set of babyChain using the validator set updates V.
+  We assume that each validator set update is unique (e.g., each validator set update V could simply have a unique identifier).
+  - Indication: <Unbonded, babyChain, V> - indication to the parent blockchain that the validator set update V (i.e., its "effect") has unbonded on the baby blockchain.
+
+Hence, we aim to achieve the following properties:
+- *Liveness*: If the parent blockchain demands a validator set change V for the baby chain, then the validator set of the baby blockchain reflects this demand.
+- *Validator set change safety*: Suppose that the validator set of the baby blockchain changes from *V* to *V'*. Then, there exists a sequence *seq = V1, V2, ..., Vn*, where *n > 0*, of change validator set demands issued by the parent blockchain such that *apply(V, seq) = V'*.
+- *Validator set change liveness*: If there exist infinitely many <ChangeValidatorSet, babyChain, V'> issued after <ChangeValidatorSet, babyChain, V>, then <Unbonded, babyChain, V> is eventually triggered.
+
+Let us explain the validator set change liveness property.
+First, note that <Unbonded, babyChain, Vlast> is never triggered if Vlast represent the **final** validator set change demand issued by the parent blockchain (the reason is that the validator set of the baby blockchain will never change after Vlast).
+Hence, it might be enough for V not to be the last demanded change.
+Unfortunatelly, that is not the case since multiple validator set change demands might be committed in the same block as the last demand.
+That is why we demand infinitely many demands to be issued after V.
+Note that if B represents the number of transactions committed in each block, then it is sufficient to demand B validator set change demands to be issued in order for V to eventually unbond.
+
+## High-level design of the Cross-Chain Validation protocol
+
+In this section, we provide an intuition behind our protocol.
+
+We use the causally consistent FIFO queue for "communication" among two blockchains (parent and baby).
```

Comment:
Isn't this an ordered channel?
Can't we stick with the existing IBC primitives in all cases we can, and if they are not sufficient, then a clear explanation of why we cannot use them, and proposals for what is needed (do you really need a DAG with causal ordering? it seems FIFO is same as "ordered")

Reply:
I was thinking about that a lot recently. For me, the write-up of the ICS focuses about send/receive packet (also when we consider how token transfer is designed/implemented), rather than on, e.g., how to read (over iterate over) e.g., all (un)acknowledged packets. I believe that these reads may be an important programming abstractions that are somehow written between the lines in the ICSs. In writing this spec, it turned out that this abstractions are quite useful. 

It might well be that our queues coincide with what can be obtained from reading the channel, as you suggest. The point for me is that we should try to make this explicit. I now believe this can be done with the existing primitives. What is missing for me is an explicit explanation/API what can be done with these primitives in addition to send/receive of individual packet.

For instance, we could introduce a function `pendingpackets()`. It returns a list of all pending packets in a channel (we can define it over the store). The return value of these function might have the same semantics as our queues. Perhaps this function already exists in the implementation? Then it would just be a matter of mentioning it in e.g., the channel ICS, and we can the use it directly for the valset update instead of introducing a new module.

Does that make sense?